packages/adapters-library/src/core/unwrapCache.ts

```diff
@@ -0,0 +1,89 @@
+import type {
+  IPricesAdapter,
+  PricesInput,
+} from '../adapters/prices-v2/products/usd/pricesV2UsdAdapter'
+import type { IProtocolAdapter } from '../types/IProtocolAdapter'
+import type { UnwrapExchangeRate, UnwrapInput } from '../types/adapter'
+import type { Chain } from './constants/chains'
+
+const TEN_MINUTES_IN_MS = 10 * 60 * 1000
+
+export interface IUnwrapCache {
+  fetchUnwrapWithCache(
+    adapter: IProtocolAdapter,
+    input: UnwrapInput,
+  ): Promise<UnwrapExchangeRate>
+  fetchPriceWithCache(
+    adapter: IPricesAdapter,
+    input: PricesInput,
+  ): Promise<UnwrapExchangeRate>
+}
+
+export class MemoryUnwrapCache implements IUnwrapCache {
+  private readonly cache: Map<
+    string,
+    Promise<{ value: UnwrapExchangeRate; expiryTime: number }>
```

Comment:
Stores a promise, not the value itself, that way it prevents multiple requests for the same value when the first one is still resolving.